### ISSUE TITLE
Now Playing Updates delayed +- 45-90 seconds

### ISSUE THREAD

**Issue by carlojou** [AzuraCast/AzuraCast#6426]
### Installation Method

Docker Installation

### AzuraCast Release Channel

Stable Channel

### Current AzuraCast Version

0.18.5

### What happened?

Hello!

We run a pretty sizable AzuraCast installation, although the node itself has tons of resources to handle it. (CPU average load is around 25-30%, RAM sitting at around ~40% used) 

However, no matter what we try - the now playing updates seem to be extremely slower than what they should be. Song updates take anywhere from 30 to 100+ seconds to update. I would suspect it's the way it's processing through each station, but not sure if there's anything we can add to correct it.

We have tried the High Performances updates option, but that yields the same results (now playing updates are still delayed by 30-100 seconds) which is odd since this should apparently address the issue of having several stations on a single instance, so perhaps our AzuraCast instance is not using that option even when it's enabled? Not sure how to check if it's actually running.

Let me know if there's any logs I can copy/paste that would be useful to have a look at. We also added the optimization tips provided here:

https://docs.azuracast.com/en/getting-started/installation/optimizing

(Also posted via Discord a while ago, but thought it would be more helpful to post it here)

Thank you!

**Comment by Vaalyn** [comment]
> We have tried the High Performances updates option, but that yields the same results (now playing updates are still delayed by 30-100 seconds) which is odd since this should apparently address the issue of having several stations on a single instance, so perhaps our AzuraCast instance is not using that option even when it's enabled? Not sure how to check if it's actually running.

If I'm not mistaking this for something else you are talking about the `Use High-Performance Now Playing Updates`, right?
This option improves latency between the UI and the server but not the latency of the server when fetching the currently now playing data for each station. There is also no mention about this improving things with many stations, just about improving the performance for large listener volumes.

See:
> Uses either Websockets, Server-Sent Events (SSE) or static JSON files to serve Now Playing data on public pages. This improves performance, especially with large listener volume. Disable this if you are encountering problems with the service or use multiple URLs to serve your public pages.

This is pretty vague:
> We run a pretty sizable AzuraCast installation

About how many stations are we talking about exactly, 10, 20, 50, 100?

The most likely culprit is probably the amount of stations and the now playing update process on the server not being able to process all of them in a short enough time.

**Comment by carlojou** [comment]
> If I'm not mistaking this for something else you are talking about the Use High-Performance Now Playing Updates, right?
This option improves latency between the UI and the server but not the latency of the server when fetching the currently now playing data for each station. There is also no mention about this improving things with many stations, just about improving the performance for large listener volumes.

That is correct, I was referring to that option. I did think it was to benefit instances with a larger station count, but I do now see it references the amount of listeners in your snippet, not the stations. 

> About how many stations are we talking about exactly, 10, 20, 50, 100?

> The most likely culprit is probably the amount of stations and the now playing update process on the server not being able to process all of them in a short enough time.

This particular instance has 120+ Stations, we started to see now-playing updates slow drastically around the 60-80 mark. The server itself is completely fine, so the bottleneck would be how the process is implemented - I've read in the past that station now-playing updates used to work in parallel in previous versions (years ago), not sure how accurate that is but I would imagine running it in parallel or in bigger batches for systems that can handle it would be a lot better than just a single process looping through each of them.

**Comment by Vaalyn** [comment]
It is still a process that does the sync in parallel but it was limited to a max. amount that are processed at the same time because it was occurring far more often that operators of installations reported CPU overload issues that were traced back to the amount of parallel now playing updates.

Currently this max. amount seems to not be configurable though.
I think it would make sense to introduce a new env var that lets server admins override the default that was set [here](https://github.com/AzuraCast/AzuraCast/blob/main/src/Console/Command/Sync/NowPlayingCommand.php#L28). 🤔 

I'm probably going to see if I can find some time on the weekend to add such an env var.

**Comment by jpcastberg** [comment]
I am seeing similar behavior - though in my case I'm only running a single station. In the video below you can see a song finish, after which it takes about 20 seconds to display the currently playing song. I also have the "Use High-Performance Now Playing Updates" setting turned on.

https://github.com/AzuraCast/AzuraCast/assets/37011196/64dc8f52-9036-4e55-b546-f677350ec2f4

**Comment by Vaalyn** [comment]
@carlojou I found some time yesterday evening to add the new env var.

Could you try increasing the `NOW_PLAYING_MAX_CONCURRENT_PROCESSES` in the `azuracast.env` from the default `5` to some higher value? Maybe something like 10 or 20. Not sure how much this will increase your CPU usage exactly but I think it should be fine since your average seems to be quite low.

Just search for the `# NOW_PLAYING_MAX_CONCURRENT_PROCESSES=5` line in the `azuracast.env`, remove the `# `, replace the number, save and then restart AzuraCast.

@jpcastberg I think in your case the source should be something different than what's happening on the installation of @carlojou since you only have a single station.

How does your CPU usage look like?
Could you post a link to your stations public page so I can take a quick look to see if I can notice anything from there?

**Comment by scysys** [comment]
> After updating to the latest Rolling-Release, could you try increasing the `NOW_PLAYING_MAX_CONCURRENT_PROCESSES` in the `azuracast.env` from the default `5` to some higher value? Maybe something like 10 or 20. Not sure how much this will increase your CPU usage exactly but I think it should be fine since your average seems to be quite low.

Thats nice. Also for other way like 1 as value :p

**Comment by jpcastberg** [comment]
@Vaalyn here is the link to the public page: https://radio.castberg.media/public/molten_death.exe

cpu usage is hovering around 10%, no issues there

thank you so much for your attention to this! let me know if you need any more info.

**Comment by Vaalyn** [comment]
@jpcastberg I've looked at the public page and at the same time at the data from `/api/nowplaying` and can confirm that in your case this is not due to the now playing update between backend process but something between the public page / the browser and the server.

I have noticed that you seem to be using CloudFlare for your station.

It seems like CF does some kind of caching even for SSE (Server Sent Events) which is used for the high performance now playing updates which is probably what's causing this delay for your public page.

There is a CF community thread I found regarding SSEs and caching where someone recommends adding a rule to exclude SSE endpoints from the CF cache to resolve such problems: https://community.cloudflare.com/t/server-sent-events-buffering/179526

Since I'm not using CF myself I can't really help you with getting that set up but I you'd need to exclude `https://radio.castberg.media/api/live/nowplaying/sse` from the caching.

**Comment by jpcastberg** [comment]
@Vaalyn sounds good 👍👍. if i get it fixed i'll open a pr to add it to the documentation.

**Comment by carlojou** [comment]
> @carlojou I found some time yesterday evening to add the new env var.
> 
> After updating to the latest Rolling-Release, could you try increasing the `NOW_PLAYING_MAX_CONCURRENT_PROCESSES` in the `azuracast.env` from the default `5` to some higher value? Maybe something like 10 or 20. Not sure how much this will increase your CPU usage exactly but I think it should be fine since your average seems to be quite low.
> 
> Just search for the `# NOW_PLAYING_MAX_CONCURRENT_PROCESSES=5` line in the `azuracast.env`, remove the `# `, replace the number, save and then restart AzuraCast.

Thank you so much! Appreciate the fast turnaround on this.

I will need to wait for the stable release, but as soon as that happens ill test it out for you and report back

**Comment by eranwo** [comment]
@Vaalyn First of all, thanks for that feature i'm looking for it for a very long time as i'm stuck in ver. v0.14.1 because of the "now palying"...
I first set the `"NOW_PLAYING_MAX_CONCURRENT_PROCESSES" `to 20 but didn't see any change, so I tried to set it to 100 but still i'm not sure.. somtimes now playing updates after 15 sec and somtimes after 45 sec. (I have 50 stations on the server)

**Comment by Vaalyn** [comment]
> @Vaalyn First of all, thanks for that feature i'm looking for it for a very long time as i'm stuck in ver. v0.14.1 because of the "now palying"... I first set the `"NOW_PLAYING_MAX_CONCURRENT_PROCESSES" `to 20 but didn't see any change, so I tried to set it to 100 but still i'm not sure.. somtimes now playing updates after 15 sec and somtimes after 45 sec. (I have 50 stations on the server)

Not sure if I understood you correctly, did you set this env var on your old 0.14.1 installation?

In order to use it you need to update to the latest Rolling-Release (or when a new stable comes out to that).

Just setting this new env var in the file without updating will not do anything, also you need to completely restart AzuraCast for it to take effect. Like `docker compose down && docker compose up -d`

**Comment by eranwo** [comment]
@Vaalyn No, sorry for the lack of clarity ... my prodaction server is on ver. v0.14.1.
I have another server for testing on the latest Rolling Release #bc5877c. And yes I restarted AzuraCast after eache change I made.

**Comment by Vaalyn** [comment]
@eranwo Can you send me a link to one of your stations public page from that test server were you have this issue and while the new env var is active?

**Comment by eranwo** [comment]
> @eranwo Can you send me a link to one of your stations public page from that test server were you have this issue and while the new env var is active?
https://aztest.livecdn.click/public/c08
NOW_PLAYING_DELAY_TIME=5
NOW_PLAYING_MAX_CONCURRENT_PROCESSES=100

**Comment by Vaalyn** [comment]
@eranwo remove the `NOW_PLAYING_DELAY_TIME=5`, this is counterproductive when you want fast updates and have enough server resources.

For reference see the description of that env var:
```
# The delay between Now Playing checks for every station.
# Decrease for more frequent checks at the expense of performance;
# increase for less frequent checks but better performance (for large installations).
```

Also, you are using SSE (Server Sent Events) with CloudFlare on your Public Page so I'm giving you the same recommendation from this comment: https://github.com/AzuraCast/AzuraCast/issues/6426#issuecomment-1636042001

**Comment by eranwo** [comment]
@Vaalyn I removed the "NOW_PLAYING_DELAY_TIME=5" (+ restart AC ) and bypassed the cache for ../api/live/now playing/sse/* but didn't see any effect.

**Comment by eranwo** [comment]
https://github.com/AzuraCast/AzuraCast/actions/runs/5548483792/jobs/10131511565

**Comment by BusterNeece** [comment]
@eranwo That was fixed in the following commit and we've had successful deploys since then.

**Comment by eranwo** [comment]
@BusterNeece Thanks. still I don't see any effect of this. Even when I remove the `NOW_PLAYING_MAX_CONCURRENT_PROCESSES` or set it to 500, the CPU resource does not change.

**Comment by eranwo** [comment]
@Vaalyn When I set "NOW_PLAYING_MAX_CONCURRENT_PROCESSES" to 100, does that mean 100 stations will be updated at the same time?

**Comment by eranwo** [comment]
@Vaalyn Please reply. Could you check it as i'm not sure it's functioning.
Thanks in advance for the effort!

**Comment by Vaalyn** [comment]
I'm not sure yet what's going on here and will need to look at it more in-depth. It should work as far as I can tell from the things I've looked at but as seen on your instance something seems to be not quite right or there is something I'm overlooking.

Currently I don't have enough time on hand so I can't exactly give an eta. as to when I'll get to look into it but I have it on my list of things to look into.

**Comment by eranwo** [comment]
@Vaalyn Hi, Have you had a chance to take a look at it?

**Comment by carlojou** [comment]
Hello!

So we've been testing this out for a couple of weeks now on the new stable release version, I can also confirm it's still not working. Results are basically exactly the same even if we increase that new setting/variable that was added (NOW_PLAYING_MAX_CONCURRENT_PROCESSES), we also tried playing around with the values to see if it would make any difference but with no luck.

Just thought I would add my two cents

**Comment by inbx** [comment]
Hi, 
I have the same issue on random stations.

• v0.19.1 Stable
• 5+ stations

**Comment by eranwo** [comment]
Hi @Vaalyn, This issue still exists in Rolling Release #1657f22 (2023-11-29 4:11). Any chance to solve it?

**Comment by Fridgemagnet** [comment]
I can confirm that there is an issue in the latest release I am on (Rolling Release #[f9f2530](https://github.com/AzuraCast/AzuraCast/commit/f9f25307a6cb526a5449920cbd4139591271d44d)). Inspecting the webpage with firefox I can see that it takes 2.5s for the info from the api/live/nowplaying endpoint for a station to load and until it does it say "station offline".
Occasionally it load instantly (normally) but more often than not there is a delay.
I have "NOW_PLAYING_DELAY_TIME=1" set and testing with "NOW_PLAYING_MAX_CONCURRENT_PROCESSES=20" does not resolve the slow now playing info load times.

**Comment by BusterNeece** [comment]
@Fridgemagnet I can't reproduce that error at the moment...Now Playing updates should be immediate upon connection to the SSE source (the `/api/live/nowplaying` URL) and on my test installations that is the case...I'd be curious to see what your overall system load characteristics look like if that's taking 2.5 seconds to connect and produce an initial result.

**Comment by Fridgemagnet** [comment]
@BusterNeece thank you for the prompt response

I solved my issue by adding the following to the nginxproxymanager proxy host for azuracast ("advanced nginx custom configuration" tab) after figuring out that the issue only occurs when accessing the radio via nginxproxymanager:

> proxy_buffering off;
> proxy_no_cache 1;
> proxy_cache_bypass 1;
> proxy_cache off;

**Comment by BusterNeece** [comment]
@Fridgemagnet You may also want to update to the latest Rolling Release, as we've reverted some changes we made that were likely impacting the Now Playing updates, especially if you had "High-Performance Now Playing Updates" enabled.

**Comment by Fridgemagnet** [comment]
That does make sense because the bug just suddenly appeared after an update.
I have now updated to the latest rolling release, thanks!

**Comment by BusterNeece** [comment]
Closing this issue for now as there's some confusion about whether new issues with Now Playing updates should be posted here (even though this issue is specifically referring to delays in receiving NP updates on a very large installation, not universal delays across the board caused by a specific bug).

If you're still encountering issues with the latest Rolling Release version delaying Now Playing updates (and by that, I mean significantly delayed, not a few seconds off the mark, as we simply can't perfectly predict the delays involved with Liquidsoap and Icecast and may miss the mark on that by a few seconds), feel free to open a new issue with the specific details of your situation. Thank you!
